src/mlpack/core/tree/binary_space_tree/rp_tree_max_split_impl.hpp

```diff
@@ -0,0 +1,114 @@
+/**
+ * @file rp_tree_max_split_impl.hpp
+ * @author Mikhail Lozhnikov
+ *
+ * Implementation of class (RPTreeMaxSplit) to split a binary space partition
+ * tree.
+ */
+#ifndef MLPACK_CORE_TREE_BINARY_SPACE_TREE_RP_TREE_MAX_SPLIT_IMPL_HPP
+#define MLPACK_CORE_TREE_BINARY_SPACE_TREE_RP_TREE_MAX_SPLIT_IMPL_HPP
+
+#include "rp_tree_max_split.hpp"
+#include "rp_tree_mean_split.hpp"
+
+namespace mlpack {
+namespace tree {
+
+template<typename BoundType, typename MatType>
+bool RPTreeMaxSplit<BoundType, MatType>::SplitNode(const BoundType& /* bound */,
+                                                  MatType& data,
+                                                  const size_t begin,
+                                                  const size_t count,
+                                                  SplitInfo& splitInfo)
+{
+  splitInfo.direction.zeros(data.n_rows);
+
+  // Get the normal to the hyperplane.
+  RPTreeMeanSplit<BoundType, MatType>::GetRandomDirection(
+      splitInfo.direction);
+
+  // Get the value according to which we will perform the split.
+  if (!GetSplitVal(data, begin, count, splitInfo.direction, splitInfo.splitVal))
+    return false;
+
+  return true;
+}
+
+template<typename BoundType, typename MatType>
+typename MatType::elem_type RPTreeMaxSplit<BoundType, MatType>::
+GetRandomDeviation(const MatType& data,
+                   const size_t begin,
+                   const size_t count,
+                   const arma::Col<ElemType>& direction)
```

Comment:
Yeah, I substituted that by 
`math::Random((minimum - splitVal) * 0.75, + (maximum - splitVal) * 0.75);`
since I am not sure that the original algorithm guarantees that both children will not be empty.